// angular-tinymce/modules/angular-tinymce.js
import { Component, EventEmitter, Inject, InjectionToken, Input, NgModule, NgZone, Output, ViewChild, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/skipWhile';
import 'rxjs/add/operator/take';
import 'rxjs/add/observable/interval';
import 'rxjs/add/observable/fromEvent';
import { CommonModule } from '@angular/common';

class TinyMceEvents {
}
TinyMceEvents.Click = 'click';
TinyMceEvents.DblClick = 'dblClick';
TinyMceEvents.MouseDown = 'MouseDown';
TinyMceEvents.MouseUp = 'MouseUp';
TinyMceEvents.MouseMove = 'MouseMove';
TinyMceEvents.MouseOver = 'MouseOver';
TinyMceEvents.MouseOut = 'MouseOut';
TinyMceEvents.MouseEnter = 'MouseEnter';
TinyMceEvents.MouseLeave = 'MouseLeave';
TinyMceEvents.KeyDown = 'KeyDown';
TinyMceEvents.KeyPress = 'KeyPress';
TinyMceEvents.KeyUp = 'KeyUp';
TinyMceEvents.ContextMenu = 'ContextMenu';
TinyMceEvents.Paste = 'Paste';
TinyMceEvents.Init = 'Init';
TinyMceEvents.Focus = 'Focus';
TinyMceEvents.Blur = 'Blur';
TinyMceEvents.BeforeSetContent = 'BeforeSetContent';
TinyMceEvents.SetContent = 'SetContent';
TinyMceEvents.GetContent = 'GetContent';
TinyMceEvents.PreProcess = 'PreProcess';
TinyMceEvents.PostProcess = 'PostProcess';
TinyMceEvents.NodeChange = 'NodeChange';
TinyMceEvents.Undo = 'Undo';
TinyMceEvents.Redo = 'Redo';
TinyMceEvents.Change = 'Change';
TinyMceEvents.Dirty = 'Dirty';
TinyMceEvents.Remove = 'Remove';
TinyMceEvents.ExecCommand = 'ExecCommand';
TinyMceEvents.PastePreProcess = 'PastePreProcess';
TinyMceEvents.PastePostProcess = 'PastePostProcess';
TinyMceEvents.AddEditor = 'AddEditor';
TinyMceEvents.RemoveEditor = 'RemoveEditor';

const TINYMCE_SETTINGS_TOKEN = new InjectionToken('angular-tinymce-settings');
class TinyMceComponent {
    /**
     * @param {?} _settings
     * @param {?} ngZone
     */
    constructor(_settings, ngZone) {
        this._settings = _settings;
        this.ngZone = ngZone;
        this.onModelChange = () => { };
        this.onModelTouched = () => { };
        this.click = new EventEmitter();
        this.dblclick = new EventEmitter();
        this.mousedown = new EventEmitter();
        this.mouseup = new EventEmitter();
        this.mousemove = new EventEmitter();
        this.mouseover = new EventEmitter();
        this.mouseout = new EventEmitter();
        this.mouseenter = new EventEmitter();
        this.mouseleave = new EventEmitter();
        this.keydown = new EventEmitter();
        this.keypress = new EventEmitter();
        this.keyup = new EventEmitter();
        this.contextmenu = new EventEmitter();
        this.paste = new EventEmitter();
        this.init = new EventEmitter();
        this.focus = new EventEmitter();
        this.blur = new EventEmitter();
        this.beforesetcontent = new EventEmitter();
        this.setcontent = new EventEmitter();
        this.getcontent = new EventEmitter();
        this.preprocess = new EventEmitter();
        this.postprocess = new EventEmitter();
        this.nodechange = new EventEmitter();
        this.undo = new EventEmitter();
        this.redo = new EventEmitter();
        this.change = new EventEmitter();
        this.dirty = new EventEmitter();
        this.remove = new EventEmitter();
        this.execcommand = new EventEmitter();
        this.pastepreprocess = new EventEmitter();
        this.pastepostprocess = new EventEmitter();
        const localSettings = this.settings || _settings || {};
        this.settings = Object.assign({}, localSettings);
        if (!window.tinymce && !document.getElementById('tinymceScript')) {
            const tag = document.createElement('script');
            tag.id = 'tinymceScript';
            tag.setAttribute('src', this.settings.tinymceScriptURL || 'assets/tinymce/tinymce.min.js');
            tag.onload = () => {
                tinymce.baseURL = this.settings.baseURL;
            };
            document.body.appendChild(tag);
        }
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        const /** @type {?} */ val = obj != null ? obj.toString() : '';
        if (this.editor) {
            this.editor.setContent(val);
        }
        else {
            this.beforeInitValue = val;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onModelChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onModelTouched = fn;
    }
    /**
     * @param {?} disabled
     * @return {?}
     */
    setDisabledState(disabled) {
        this.disabled = disabled;
        this.setEditorMode(disabled);
    }
    /**
     * @param {?} disabled
     * @return {?}
     */
    setEditorMode(disabled) {
        if (this.editor) {
            if (disabled) {
                this.editor.setMode('readonly');
            }
            else {
                this.editor.setMode('design');
            }
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['settings']) {
            this.settings = Object.assign({}, changes['settings'].currentValue);
        }
        if (changes['isDisabled']) {
            this.setDisabledState(changes['isDisabled'].currentValue);
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.settings.target = this.elem.nativeElement;
        this.initCallbacks(this.settings);
        Observable
            .interval(300)
            .skipWhile(() => !((window)).tinymce)
            .take(1)
            .subscribe(() => {
            tinymce.init(this.settings);
        });
    }
    /**
     * @param {?} settings
     * @return {?}
     */
    initCallbacks(settings) {
        const /** @type {?} */ orig = settings.init_instance_callback;
        settings.init_instance_callback = (editor) => {
            this.editor = editor;
            this.setEditorMode(this.disabled);
            if (this.beforeInitValue != null) {
                this.editor.setContent(this.beforeInitValue);
            }
            if (orig) {
                orig(editor);
            }
            editor.on(TinyMceEvents.Click, (e) => this.click.emit(e));
            editor.on(TinyMceEvents.DblClick, (e) => this.dblclick.emit(e));
            editor.on(TinyMceEvents.MouseDown, (e) => this.mousedown.emit(e));
            editor.on(TinyMceEvents.MouseUp, (e) => this.mouseup.emit(e));
            editor.on(TinyMceEvents.MouseMove, (e) => this.mousemove.emit(e));
            editor.on(TinyMceEvents.MouseOver, (e) => this.mouseover.emit(e));
            editor.on(TinyMceEvents.MouseOut, (e) => this.mouseout.emit(e));
            editor.on(TinyMceEvents.MouseEnter, (e) => this.mouseenter.emit(e));
            editor.on(TinyMceEvents.MouseLeave, (e) => this.mouseleave.emit(e));
            editor.on(TinyMceEvents.KeyDown, (e) => this.keydown.emit(e));
            editor.on(TinyMceEvents.KeyPress, (e) => this.keypress.emit(e));
            editor.on(TinyMceEvents.KeyUp, (e) => {
                this.ngZone.run(() => {
                    this.onModelChange(this.editor.getContent());
                    this.onModelTouched();
                });
                this.keyup.emit(e);
            });
            editor.on(TinyMceEvents.ContextMenu, (e) => this.contextmenu.emit(e));
            editor.on(TinyMceEvents.Paste, (e) => this.paste.emit(e));
            editor.on(TinyMceEvents.Focus, (e) => this.focus.emit(e));
            editor.on(TinyMceEvents.Blur, (e) => this.blur.emit(e));
            editor.on(TinyMceEvents.BeforeSetContent, (e) => this.beforesetcontent.emit(e));
            editor.on(TinyMceEvents.SetContent, (e) => this.setcontent.emit(e));
            editor.on(TinyMceEvents.GetContent, (e) => this.getcontent.emit(e));
            editor.on(TinyMceEvents.PreProcess, (e) => this.preprocess.emit(e));
            editor.on(TinyMceEvents.PostProcess, (e) => this.postprocess.emit(e));
            editor.on(TinyMceEvents.NodeChange, (e) => this.nodechange.emit(e));
            editor.on(TinyMceEvents.Undo, (e) => this.undo.emit(e));
            editor.on(TinyMceEvents.Redo, (e) => this.redo.emit(e));
            editor.on(TinyMceEvents.Change, (e) => this.change.emit(e));
            editor.on(TinyMceEvents.Dirty, (e) => this.dirty.emit(e));
            editor.on(TinyMceEvents.Remove, (e) => this.remove.emit(e));
            editor.on(TinyMceEvents.ExecCommand, (e) => {
                this.ngZone.run(() => {
                    const /** @type {?} */ content = this.editor.getContent();
                    if (content != null && content.length > 0) {
                        this.onModelChange(content);
                        this.onModelTouched();
                    }
                });
                this.execcommand.emit(e);
            });
            editor.on(TinyMceEvents.PastePreProcess, (e) => this.pastepreprocess.emit(e));
            editor.on(TinyMceEvents.PastePostProcess, (e) => this.pastepostprocess.emit(e));
        };
        settings.setup = (editor) => {
            editor.on(TinyMceEvents.Init, (e) => this.init.emit(e));
        };
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.editor) {
            ((tinymce)).remove(this.editor);
        }
    }
}
TinyMceComponent.decorators = [
    { type: Component, args: [{
                selector: 'angular-tinymce',
                template: `<textarea #tinymce ></textarea>`,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => TinyMceComponent),
                        multi: true
                    }]
            },] },
];
/**
 * @nocollapse
 */
TinyMceComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TINYMCE_SETTINGS_TOKEN,] },] },
    { type: NgZone, },
];
TinyMceComponent.propDecorators = {
    'isDisabled': [{ type: Input },],
    'settings': [{ type: Input },],
    'selector': [{ type: Input },],
    'click': [{ type: Output },],
    'dblclick': [{ type: Output },],
    'mousedown': [{ type: Output },],
    'mouseup': [{ type: Output },],
    'mousemove': [{ type: Output },],
    'mouseover': [{ type: Output },],
    'mouseout': [{ type: Output },],
    'mouseenter': [{ type: Output },],
    'mouseleave': [{ type: Output },],
    'keydown': [{ type: Output },],
    'keypress': [{ type: Output },],
    'keyup': [{ type: Output },],
    'contextmenu': [{ type: Output },],
    'paste': [{ type: Output },],
    'init': [{ type: Output },],
    'focus': [{ type: Output },],
    'blur': [{ type: Output },],
    'beforesetcontent': [{ type: Output },],
    'setcontent': [{ type: Output },],
    'getcontent': [{ type: Output },],
    'preprocess': [{ type: Output },],
    'postprocess': [{ type: Output },],
    'nodechange': [{ type: Output },],
    'undo': [{ type: Output },],
    'redo': [{ type: Output },],
    'change': [{ type: Output },],
    'dirty': [{ type: Output },],
    'remove': [{ type: Output },],
    'execcommand': [{ type: Output },],
    'pastepreprocess': [{ type: Output },],
    'pastepostprocess': [{ type: Output },],
    'elem': [{ type: ViewChild, args: ['tinymce',] },],
};

/**
 * @return {?}
 */
function tinymceDefaultSettings() {
    return {
        tinymceScriptURL: '/assets/tinymce/tinymce.min.js',
        theme_url: '/assets/tinymce/themes/modern/theme.min.js',
        skin_url: '/assets/tinymce/skins/lightgray'
    };
}

/**
 * Modul für ein TinyMce Input
 */
class TinyMceModule {
    /**
     * @param {?} settings
     * @return {?}
     */
    static forRoot(settings) {
        return {
            ngModule: TinyMceModule,
            providers: [
                { provide: TINYMCE_SETTINGS_TOKEN, useValue: settings }
            ]
        };
    }
}
TinyMceModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [TinyMceComponent],
                exports: [TinyMceComponent]
            },] },
];
/**
 * @nocollapse
 */
TinyMceModule.ctorParameters = () => [];

/**
 * Angular library starter.
 * Build an Angular library compatible with AoT compilation & Tree shaking.
 * Written by Roberto Simonetti.
 * MIT license.
 * https://github.com/robisim74/angular-library-starter
 */
/**
 * Entry point for all public APIs of the package.
 */
//export * from './src/angular-tinymce';

/**
 * Generated bundle index. Do not edit.
 */

export { TinyMceComponent, TINYMCE_SETTINGS_TOKEN, tinymceDefaultSettings, TinyMceEvents, TinyMceModule };
